R/query_search.R

```diff
@@ -0,0 +1,414 @@
+           " WHERE report_version.report = $1")
+  res <- DBI::dbGetQuery(con, paste(sql, collapse = "\n"), name)
+
+  value <- as.list(res$value)
+  i <- res$type == "number"
+  value[i] <- as.numeric(res$value[i])
+  i <- res$type == "boolean"
+  value[i] <- as.logical(res$value[i])
+  names(value) <- res$name
+  split(value, res$report_version)
+}
+
+
+search_tags_archive <- function(name, config) {
+  con <- orderly_db("destination", config)
+  on.exit(DBI::dbDisconnect(con))
+
+  sql <- c("SELECT report_version_tag.*",
+           "  FROM report_version_tag",
+           "  JOIN report_version",
+           "    ON report_version.id = report_version_tag.report_version",
+           " WHERE report_version.report = $1")
+  res <- DBI::dbGetQuery(con, paste(sql, collapse = "\n"), name)
+  split(res$tag, res$report_version)
+}
+
+
+search_x_draft <- function(name, config, extract) {
+  path <- file.path(config$root, "draft", name)
+  versions <- orderly_list_dir(path)
+  f <- function(p) {
+    extract(readRDS(path_orderly_run_rds(p)))
+  }
+  set_names(lapply(file.path(path, versions), f), versions)
+}
+
+
+search_parameters_draft <- function(name, config) {
+  search_x_draft(name, config, function(x) x$meta$parameters)
+}
+
+
+search_tags_draft <- function(name, config) {
+  search_x_draft(name, config, function(x) x$meta$tags)
+}
+
+
+parse_query_operators <- c("is.null", "==", "!=", ">", ">=", "<", "<=")
+parse_query_join <- c("(", "&&", "||", "!")
+
+
+parse_query <- function(x, parameters) {
+  expr <- parse(text = x)
+  if (length(expr) != 1L) {
+    stop("Expected a single expression")
+  }
+  expr <- expr[[1L]]
+
+  latest <- is_call(expr, "latest")
+  if (latest) {
+    stopifnot(length(expr) == 2L)
+    expr <- expr[[2L]]
+  }
+
+  dat <- parse_query_expr(expr, parameters)
+
+  test <- c("parameter", "tag")
+  use <- set_names(as.list(test %in% dat$type), test)
+  expr <- dat$expr
+  list(latest = latest, use = use, expr = expr)
+}
+
+
+parse_query_expr <- function(expr, parameters) {
+  if (!is.recursive(expr)) {
+    stop("Invalid expression!")
+  }
+
+  if (is.call(expr) && deparse(expr[[1L]]) %in% parse_query_join) {
+    fn <- deparse(expr[[1]])
+
+    a <- parse_query_expr(expr[[2]], parameters)
+    expr[[2]] <- a$expr
+
+    if (fn %in% c("&&", "||")) {
+      b <- parse_query_expr(expr[[3]], parameters)
+      expr[[3]] <- b$expr
+    } else {
+      b <- NULL
+    }
+
+    list(type = c(a$type, b$type), expr = expr)
+  } else {
+    parse_query_filter(expr, parameters)
+  }
+}
+
+## examples
+##
+## * id == 1
+## * parameter:id == 1
+## * tag:weekly
+parse_query_filter <- function(expr, parameters) {
+  if (!is.call(expr)) {
+    stop("Expected an expression")
+  }
+
+  if (is_call(expr, ":") && length(expr) == 3L) {
+    type <- expr[[2L]]
+    stopifnot(is.symbol(type))
+    if (!identical(type, quote(tag))) {
+      stop(sprintf("Invalid query expression '%s' requires operator",
```

Comment:
I think this will be clearer about what the problem is if we pass the `type` to the error message.

Something like
`Invalid query expression 'z:thing', operator 'z' not supported - supported operators: 'tag', 'parameter'`

Reply:
Ah I see this captures when you pass `paramter:thing` in too without an operator like `>`.

What I am trying to say is if we do something like
```
orderly::orderly_search("something:123", "other", root = root)
```
We get an error

```
 Error: Invalid query expression 'test:123' requires operator
```
which is a bit misleading. I can see people trying to do that with some other search term other than `tag` or `parameter` but perhaps that is unlikely to happen?

Reply:
Thanks - this was very useful. Have updated now so errors are like

```
> orderly::orderly_search("something:123", "other", root = path)
Error: Invalid namespaced query element 'something:123'; expected symbol for key
> orderly::orderly_search("something:abc", "other", root = path)
Error: Query namespace (used as 'something') must be one of 'tag', 'parameter'
```